### PR TITLE
[IoT - Device Provisioning] DataPlane API Review

### LINKED ISSUE #15694: [DPS] REST API review for Azure IoT Device Provisioning Service (DPS)
- Service name: Azure IoT Device Provisioning Service (DPS)
- Key contact for the review: @lerubin, @anushr
- Whether this is a new or existing API: existing
- Whether or not your service already been deployed: already deployed
- A brief description of the material to be reviewed/changed: reviewing general conformity of the Data Plane service API to the Azure API Guidelines plus alignment with other IoT Services
- A link to the PR/issue in the azure-rest-api-specs (public)  or azure/azure-rest-api-specs-pr (private) repo:
  - [DPS DataPlane API Review by leigharubin · Pull Request #15621](https://github.com/Azure/azure-rest-api-specs/pull/15621)

### PR DESCRIPTION
**Pull request by leigharubin**
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - Onboarding our Data Plane swagger to the **azure-rest-api-specs** repo. We are already a GA service (have been for several years), so this PR is our next API version but only the first one being merged to this repo. Thus, there will be make suggested changes which we can not fix do to needing to avoid breaking changes.
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  - October 2021
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [x] SDK of .NET (need service team to ensure code readiness)
      - [x] SDK of Python
      - [x] SDK of Java
      - [x] SDK of Js
      - [x] SDK of Go
      - [x] PowerShell
      - [x] CLI
      - [?] Terraform

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that label “WaitForARMFeedback” will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits.

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (mikekistler) [inline specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:75]: You should specify a `maxLength` and `pattern` on path parameters to restrict the length and allowed set of characters for user-specified names, e.g.
```suggestion
            "type": "string",
            "maxLength": 50,
            "pattern": "^[a-z0-9-]+$",
```

**Review by mikekistler — CHANGES_REQUESTED** [review]
There are a number of changes needed -- some minor but others more fundamental, like using the Azure error response body schema and pagination pattern.

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:1]: Hi @mikekistler , DPS has had many API versions prior to this one but they have not been uploaded to this repo. We have been GA for several years now but are just uploading our most recent API version here, hence why it is not preview.

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:52]: Ack, will have to keep this as is to avoid breaking change.

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:428]: Ack, will keep in favor of no breaking changes

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:44]: Ack, this is currently in response body, will keep in body to avoid breaking change. Would there be a benefit of adding in the header as well?

Comment (mikekistler) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:44]: Sorry — I missed the fact that it was already returned in the response body. In that case pls disregard my comment.

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:397]: Ack, will leave to avoid breaking change.

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:390]: Ack, will keep to avoid breaking change

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:430]: This is essentially a list of all the items returned in the query (which can be DeviceRegistrationStates, Enrollments, EnrollmentGroups based on the API being used. Do you have examples of what other services have used for this type of item type? Was thinking along the lines of "The list of values returned from the query."

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:782]: Ack, keeping to avoid breaking changes

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:397]: Ack, keeping to avoid breaking changes

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (mikekistler) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:75]: Good question! I hope that client SDKs do not validate this but I will confirm.

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:1283]: @mikekistler does "maxLength" work here? This is what our current Swashbuckle version generates, looks like this was fixed to add "maxItems" in v5.6.0 https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/1747

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:34]: This registration ID refers to the same ID as the registration ID in DeviceRegistration object. Will update the description to make this clear.

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (mikekistler) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:1283]: “maxLength” is for strings — you want “maxItems” here.  If you can’t get “maxItems” with your current tooling and don’t want to update, then just leave it out.  It really only benefits the documentation.

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:1057]: All properties in the code have a description, but only some generate in the swagger due to the version of Swashbuckle we are using. Will create a PBI to update Swashbuckle, but this is a large task since the next version contains many breaking changes + a new object model. Closing this item for now.

PBI: https://dev.azure.com/msazure/One/_workitems/edit/10911796

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json]: Added

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (leigharubin) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json]: @mikekistler I have added the regex here, please confirm that SDKs do not validate and I will resolve this.

**Comment by leigharubin** [comment]
@mikekistler is there a format / template for the README that is blocking this PR from merging?

**Comment by mikekistler** [comment]
I found a sample readme file here:

https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/92/Sample-ReadME

I think (hope) that is what you need.

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (mikekistler) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json]: @leigharubin I looked into the code generated by autorest for a path parameter with a `pattern`, and I was surprised to see that the client does have logic to validate the pattern.  It’s not clear exactly when this logic is used, but the fact that it is there at all makes me second guess whether adding `pattern` here is a good idea.  It’s probably best to drop it for now.  Sorry …

**Review by mikekistler — COMMENTED** [review]
All my concerns have been addressed. 👍

**Comment by mikekistler** [comment]
@JeffreyRichter @markweitzel @johanste @tg-msft @rysweet This API was through the board and the notes are in #15694. I think the team has addressed the whatever issues they could without introducing breaking changes (the service is already GA -- this PR is  just to get the API defn into the repo).

Could we get one more thumbs up on this PR before we move it to approved?

**Review by markweitzel — APPROVED** [review]
Scope of review is captured in #15694 
Looks good. Thank you.

**Review by JeffreyRichter — APPROVED** [review]
I wish the swagger was organized by HTTP methods (PUT, POST, GET, etc.). It should be organized by resource type instead.

**Comment by leigharubin** [comment]
> Scope of review is captured in #15694 Looks good. Thank you.

@markweitzel Thank you! As noted in the PR, we will not be able to fix the "Swagger LintDiff" errors as this is a GA service but our first time uploading our swagger here, so fixing this would cause breaking changes. Could you please merge this PR without us needing to fix that automated check? @mikekistler said you could potentially help with this. Thanks!

**Comment by markweitzel** [comment]
> > Scope of review is captured in #15694 Looks good. Thank you.
> 
> @markweitzel Thank you! As noted in the PR, we will not be able to fix the "Swagger LintDiff" errors as this is a GA service but our first time uploading our swagger here, so fixing this would cause breaking changes. Could you please merge this PR without us needing to fix that automated check? @mikekistler said you could potentially help with this. Thanks!

@leigharubin -- I don't have perms to do the merge but will ping @lmazuel who, I believe does have permission.

**Comment by tjprescott** [comment]
@mikekistler if you are satisfied with the changes can you change your review to an approval? We should then be good to merge.

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (mikekistler) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:887]: Schema names are not part of the REST API contract.  This change would only be breaking if you have generated clients from this API definition _and_ you intend to update those generated clients to support this new version.

Is that the case?

**Review by mikekistler — COMMENTED** [review]
I am generally good with this except for my concern about simpler names.  If we are going to create Azure SDK libraries for this service -- that appears to be the plan -- then this is our best opportunity to get the names right.

**Comment by leigharubin** [comment]
> I am generally good with this except for my concern about simpler names. If we are going to create Azure SDK libraries for this service -- that appears to be the plan -- then this is our best opportunity to get the names right.

@mikekistler unfortunately due to needing to maintain back-compat when upgrading API versions, we will not be able to change these names as they will cause friction for the customer to upgrade to use new features.

**Comment by tjprescott** [comment]
@leigharubin I don't think @mikekistler means changing the names in the service, but aliasing them for SDK generation with `x-ms-client-name`. If this would be your first generated SDK, that would be prudent.

**Comment by leigharubin** [comment]
> @leigharubin I don't think @mikekistler means changing the names in the service, but aliasing them for SDK generation with `x-ms-client-name`. If this would be your first generated SDK, that would be prudent.

@tjprescott I see, I was misunderstanding at first. However, we already have many versions of SDKs generated so I believe the same issue applies here as well.

File: specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json
Comment (mikekistler) [reply specification/deviceprovisioningservices/data-plane/Microsoft.Devices/stable/2021-10-01/service.json:887]: Resolved offline. No need to make this change.

**Review by mikekistler — APPROVED** [review]
Looks good! 👍 

All my concerns have been addressed.

**Pull request merged**
